flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/restore/ChangelogRestoreTarget.java

```diff
@@ -21,52 +21,143 @@
 
     /** Returns the key groups which this restore procedure covers. */
-    KeyGroupRange getKeyGroupRange();
+    public KeyGroupRange getKeyGroupRange() {
+        return keyedStateBackend.getKeyGroupRange();
+    }
+
+    /**
+     * Returns the existing state created by {@link #createKeyedState(TypeSerializer,
+     * StateDescriptor)} or {@link #createPqState(String, TypeSerializer)} in the restore procedure.
+     */
+    public ChangelogState getExistingState(
+            String name, StateMetaInfoSnapshot.BackendStateType type) {
+        return changelogStateFactory.getExistingState(name, type);
+    }
 
     /**
      * Creates a keyed state which could be retrieved by {@link #getExistingState(String,
-     * BackendStateType)} in the restore procedure. The interface comes from {@link
-     * KeyedStateBackend#getOrCreateKeyedState(TypeSerializer, StateDescriptor)}.
+     * BackendStateType)} in the restore procedure.
      */
-    <N, S extends State, V> S createKeyedState(
+    @SuppressWarnings("unchecked")
+    public <N, S extends State, V> S createKeyedState(
             TypeSerializer<N> namespaceSerializer, StateDescriptor<S, V> stateDescriptor)
-            throws Exception;
+            throws Exception {
+        ChangelogState existingState =
+                changelogStateFactory.getExistingState(
+                        stateDescriptor.getName(),
+                        StateMetaInfoSnapshot.BackendStateType.KEY_VALUE);
+        if (existingState == null
+                || !isCompleteCompatible(
+                        (InternalKvState<K, N, V>) existingState,
+                        namespaceSerializer,
+                        stateDescriptor)) {
+            S keyedState =
+                    keyedStateBackend.upgradeKeyedState(namespaceSerializer, stateDescriptor);
+            functionDelegationHelper.addOrUpdate(stateDescriptor);
+            final InternalKvState<K, N, V> kvState = (InternalKvState<K, N, V>) keyedState;
+            return (S)
+                    changelogStateFactory.create(
+                            stateDescriptor,
+                            kvState,
+                            getKvStateChangeLogger(kvState, stateDescriptor),
+                            keyedStateBackend);
+        }
+        return (S) existingState;
+    }
+
+    private <N, S extends State, V> boolean isCompleteCompatible(
+            InternalKvState<K, N, V> existingState,
+            TypeSerializer<N> namespaceSerializer,
+            StateDescriptor<S, V> stateDescriptor) {
+        return isCompleteCompatible(existingState.getNamespaceSerializer(), namespaceSerializer)
+                && isCompleteCompatible(
+                        existingState.getValueSerializer(), stateDescriptor.getSerializer());
+    }
+
+    private <T> boolean isCompleteCompatible(
+            TypeSerializer<T> originalSerializer, TypeSerializer<T> newSerializer) {
+        return originalSerializer.equals(newSerializer);
```

Comment:
Is the contract of Serializers `equals` suitable for migration?
Shouldn't it be something like `originalSerializer.snapshotConfiguration().resolveSchemaCompatibility(newSerializer).isIncompatible()` ?

Reply:
I mean we could and should use this same logic from `StateSerializerProvider` (maybe extracting to a common place).
With `equals`, I'm concerned that the migration will be performed unnecessarily (or not performed at all if `equals` isn't implemented correctly).